rtos/TARGET_CORTEX/SysTimer.cpp

```diff
@@ -100,43 +102,40 @@ bool SysTimer::suspend_time_passed()
 
 uint32_t SysTimer::resume()
 {
-    core_util_critical_section_enter();
+    // Remove ensures serialized access to SysTimer by stopping timer interrupt
+    remove();
 
     _suspended = false;
     _suspend_time_passed = true;
-    remove();
 
-    uint64_t new_tick = (ticker_read_us(_ticker_data) - _start_time) * OS_TICK_FREQ / 1000000;
-    if (new_tick > _tick) {
+    uint64_t elapsed_ticks = (ticker_read_us(_ticker_data) - _time_us) / US_IN_TICK;
```

Comment:
Gah!!!  I briefly wondered whether this might be a solution, but immediately shot it down because I assumed `SysTimer::resume` was being called in a critical section from outside.  This is great.

I think it's still worth optimising for the common short case - if this is taking 20us every millisecond, that is 2% of CPU time just for the division. (On a 4MHz system I saw about 10% of CPU time being eaten up by the millisecond tickers.).

How about:

    uint64_t elapsed_us = ticker_read_us(_ticker_data) - _time_us;
    uint64_t elapsed_ticks;
    if (elapsed_us > 0xFFFFFFFF) {
        // This will typically invoke full 64/64 software division
        elapsed_ticks = elapsed_us / US_IN_TICK;
    } else {
        // This will will invoke 32/32 software division on ARMv6-M or ARMv8-M baseline, and
        // on ARMv7-M or ARMv8-M mainline can be optimised to a UMULL by 2^n/US_IN_TICK, or
        // at worst be a UDIV instruction. Thus it will be faster, maybe greatly.
        elapsed_ticks = (uint32_t) elapsed_us / US_IN_TICK;
    }

Reply:
> I assumed `SysTimer::resume` was being called in a critical section from outside.

It used to be in a critical section, but had to be changed when the RTX API it relied on became unavailable. I missed it at this time as well.

Initially I had it avoid division all together in the fast case - when only a single tick had elapsed: 
```C
if (elapsed_time < US_IN_TICK) {
    elapsed_ticks = 0;
} else if (elapsed_time < US_IN_TICK * 2) {
    elapsed_ticks = 1;
} else {
    elapsed_ticks = elapsed_time / US_IN_TICK;
}
```

I ended up going with the simpler implementation you see now since I was just worried about reducing worst case execution time. Also, I was only setup to profile only worst-case execution time, which did change much from this.

It sounds like you have profiled the average execution time on a couple of devices. Would you be able to profile the different optimizations and create a PR for the one which looks the most promising?

Reply:
I'll have to see if those extra cases help too. My feeling is that division tends to have early termination, so might not help much, but worth checking.

> I was just worried about reducing worst case execution time.

But we don't care about worst-case any more if we're outside of the critical section, right?

Reply:
> But we don't care about worst-case any more if we're outside of the critical section, right?

True, worst case doesn't matter anymore. At the time I was only setup to profile worst-case, so I wasn't able to evaluate the impact of this change.